Unify settings

blizzz opened the pull request:
Merging Personal and Admin settings into one

![spectacle h24509](https://cloud.githubusercontent.com/assets/2184312/26112455/3c7ce770-3a58-11e7-84d8-6357dd08bd35.png)

blizzz committed:
add route and controller. consolidate common settings functions in a trait.

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
one settings link, common template, styling

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
Extend interfaces

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
brief, incomplete personal info settings implementation

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
completing PersonalInfo

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
fix displaying groups in personal info

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
add storage info, fix URL generation

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
convert Sessions section

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
take over app passwords

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
sync clients have their own place

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

MorrisJobke commented:
@blizzz What is missing here?

blizzz commented:
@MorrisJobke adjusting Manager and DB (almost done, locally), fallback entry, converting other core apps, unit tests

blizzz committed:
Adopt DB and Manager to personal settings structure

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
add Additional (+ Fallback) section for  Personal settings

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
Now _converting other core apps_ and _adjust/extend unit tests_ are missing. First could be done in a separate PR,  too, as second phase. Since  they all get pulled into  "Additional".

blizzz committed:
remove obsolete files

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
add missing pieces to Settings Manager and fix and extend its unit tests

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

icewind1991 committed:
Fix trait usage

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 committed:
fix getting sections

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 commented:
- rebased
- fixed trait usage
- fixed loading sections from db

icewind1991 committed:
move personal external storage settings to it's own section

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 commented:
- moved files_external's personal settings to the new system as example (also incorporating https://github.com/nextcloud/server/pull/5173)

blizzz commented:
* [ ] settings/personal.php take over needs to be adjusted, after https://github.com/nextcloud/server/pull/5084 got merged

blizzz committed:
take over changes from  #5084

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
take over changes from #5166 and follow #5267

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
…and adjust tests

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
adjust another test

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
drone wasn't satisfied yet

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
update autoloader

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
already declared in Trait

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
Fix and extend acceptance tests

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

jancborchardt commented:
Looking good! :) Will probably adjust the headings a bit but later.

A big thing we really need to solve is the hard-refresh on navigation change. :\ It’s really bad UX and takes very long to load.

blizzz commented:
Therefore we need to fix all the javascripts, i think, because they do not survive being loaded again or initialize again. or someone comes up with a better idea.

nickvergessen commented:
@blizzz so this is good to go as a first step?

blizzz committed:
remove obsolete file

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
Missing apps form this repo as of now:
* [ ] encryption
* [ ] twofactor_backupcodes
* [ ] federatedfilesharing

blizzz committed:
adapt federatedfilesharing

also drops IE8 switch

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
do not offer additional settings when there is no content

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
adapt twofactore_backupcodes

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
adapt encryption

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
avoid marking two sections as active when they have the same name

… in both personal and admin.

blizzz commented:
OK, ready to review.

What I do not like design wise is that now "Security", "Second factor auth" and "Encryption" use all the lock icon…

![spectacle l26477](https://user-images.githubusercontent.com/2184312/27444604-275fbf0c-5777-11e7-8713-dbbca5d705c0.png)

blizzz commented:
ah, breaking tests… looking into them later.

nickvergessen commented:
I think they should use the security category. We just didn't have it yet in the personal section.

blizzz commented:
@nextcloud/designers thoughts on this? ^

MorrisJobke commented:
> @nextcloud/designers thoughts on this? ^

I would also combine the security and second factor at least.

nickvergessen commented on apps/federatedfilesharing/templates/settings-personal.php:
```diff
@@ -18,7 +18,6 @@
 
 		<br>
 
-		<?php if ($_['showShareIT']) {?>
```
accident?

blizzz replied:
by purpose. The old assignment was whether the user agent was IE8, and we dropped support since.

nickvergessen commented:
This branch has conflicts that must be resolved

blizzz commented:
also need to adapt to the force language settings 😩 time to get this in soon afterwards 😀

blizzz committed:
adapt to force language changes

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
move twofactor_* into security

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
fix setting up encryption section

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
take out additional (legacy) settings section from tests

because it is dependent whether anything else registers into it and
\OC_App is static

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
forgotten test adjustments

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
All done, final reviews por favore :) 

@schiessle especially a look on Encryption is appreciated 🔥🤚

danxuliu commented on tests/acceptance/features/bootstrap/SettingsMenuContext.php:
```diff
@@ -83,6 +93,14 @@ public function iOpenTheUserSettings() {
+	 */
+	public function iVisitTheSettingsPage() {
+		$this->iOpenTheSettingsMenu();
+		$this->actor->find(self::menuItemFor('Settings'), 20)->click();
```
May I suggest to use a timeout of 2 instead for consistency with other actions that interact with settings menu items (like "_iOpenTheUserSettings()_")?

blizzz replied:
I understand the timeout is used as max time until the settings page opens. 2 sec feel a bit short. There's just two other clicks with a 2s timeout and another one with a 10s timeout. It works and does not do any bad.

danxuliu replied:
>  I understand the timeout is used as max time until the settings page opens.

No, it is not; that timeout is the max time until the "_Settings_" item of the main "_Settings_" menu is found after the main "_Settings_" menu is opened, so 2 seconds should be enough (and in the case of the Drone server the timeouts are automatically extended, so it should be no problem even if the Drone server is loaded).

blizzz replied:
Okay, but this is about clicking on a link? Anyway, I try it, and it will either pass or not :)

danxuliu replied:
>  Okay, but this is about clicking on a link?

Yes, `find(self::menuItemFor('Settings'), XXX)->click()` waits up to XXX seconds to find the menu item named _Settings_; if it is found then it is clicked, and if it is not found an exception is thrown instead, which aborts the test.

>  Anyway, I try it, and it will either pass or not :)

I would be very surprised if it did not ;)

danxuliu reviewed (commented):
Just minor style issues from your friendly nitpicker danxuliu :P

I do not want to make you work just due to my own style preferences, so I can amend the related commit myself if you agree with that ;)

blizzz commented on tests/acceptance/features/core/ElementWrapper.php:
```diff
@@ -119,7 +119,7 @@ public function doNotHandleFailedCommands() {
 	 * Returns whether the wrapped element is visible or not.
 	 *
-	 * @return boolbean true if the wrapped element is visible, false otherwise.
+	 * @return bool true if the wrapped element is visible, false otherwise.
```
```
$ ack --type=php "@return boolean" --ignore-directory=3rdparty --ignore-directory=build  -c -h 
359
$ ack --type=php "@return (bool$|bool )" --ignore-directory=3rdparty --ignore-directory=build  -c -h 
1170
```

``bool`` wins :D conistency should to be reached within  the whole codebase, not subparts of it. Also ``bool`` is more common within PHP, for instance also when used in type hinting [[1](https://secure.php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration)].

ChristophWurst replied:
FYI https://phpdoc.org/docs/latest/references/phpdoc/types.html says

> boolean or bool, the element to which this type applies only has state true or false.

and it allows both ;-)

danxuliu replied:
Oops, my bad; I knew that "_bool_" was valid too, but I thought that "_boolean_" was a better name (that is what happens when you come from a Java background :P ); `bool` then 👍

blizzz commented on tests/acceptance/features/access-levels.feature:
```diff
@@ -14,8 +13,19 @@ Feature: access-levels
     And I see that the "Help" item in the Settings menu is shown
     And I see that the "Log out" item in the Settings menu is shown
+
+  Scenario: regular users cannot see amdin-level items on the Settings page
```
Thx for the typo!

Otherwise, it is the page where ``https://your.server.tld/settings/user`` brings  you, which is ``Personal Info`` (but this is a matter of routing). So, the default "Settings" page you reach from the menu. Definitely is a standalone page, thus I prefer to keep it this way. Or change it to "_default Settings page_" if it makes more clear.

danxuliu replied:
Yes, it is a standalone page in the sense that, once logged in, you can reach it directly by typing its URL in the address bar and it does not depend on anything else. However, with standalone I was refering to a page that you typically reach "_outside_" the application, something that you cannot reach by interacting with menus and buttons in the application and that it is not integrated with the main UI, like the login and authentication pages; that is why I was calling it "_the Settings_" instead of "_the Settings page_". Anyway, as I said I admit it is a rather arbitrary criterion, so feel free to name it as you prefer ;)

blizzz replied:
Confusing, for me a page is a page, independent of whether I need to be authenticated or not. Standalone… OK this I can understand, however with simply "page" in the Scenario name it sounds pretty indifferent to me. Looking at the other acceptance tests:

* Page within the scenario description was not used previously.

* In ``app-files.feature`` you also visit a page that is beyond  authentication ( ``Then I see that the current page is the shared link I wrote down``).

* And also in ``login.feature`` it is used when accessing a page available only when logged in: ``Then I see that the current page is the Files app``

Imho it is not a different usage. Or is it?

danxuliu replied:
>  Confusing, for me a page is a page, independent of whether I need to be authenticated or not.

Yes, yes, whether you are authenticated or not does not change if something is a page or not; I simply used the login page as an example in the sense of being "_outside_" the application; if there was, I do not know, something like _www.nextcloudserver.com/nuke_ that was accessible only to a logged in admin by typing its URL in the address bar and simply showed a button to wipe the database I would call it a page too.

>  Standalone… OK this I can understand, however with simply "page" in the Scenario name it sounds pretty indifferent to me.

Ok, then do not worry, left it as is (yes, I feel pretty bad for making you lose your time on this...); as I said it was simply a personal preference ;)

>  - In app-files.feature you also visit a page that is beyond authentication ( Then I see that the current page is the shared link I wrote down).
>  
>  - And also in login.feature it is used when accessing a page available only when logged in: Then I see that the current page is the Files app

As I said above whether the user is authenticated or not is not the criterion I use to call something a _page_. Anyway, as I said in the first comment the term _page_ in _I see that the current page is XXX_ is already inconsistent with that criterion, as in _I see that the current page is XXX_ it is being used in its broad sense, so shame on me :P

>  Imho it is not a different usage. Or is it?

Well, for me it is, but I am aware that it is a blurry concept poorly explained :P (but I tried! ;) )

blizzz commented on tests/acceptance/features/access-levels.feature:
```diff
@@ -14,8 +13,19 @@ Feature: access-levels
     And I see that the "Log out" item in the Settings menu is shown
+
+  Scenario: regular users cannot see amdin-level items on the Settings page
+    Given  I am logged in
```
✔️

blizzz commented on tests/acceptance/features/access-levels.feature:
```diff
@@ -14,8 +13,19 @@ Feature: access-levels
+    And I see that the "Administration" settings panel is not shown
+
+  Scenario: admin users can see amdin-level items on the Settings page
+    Given  I am logged in as the admin
```
✔️

blizzz commented on tests/acceptance/features/bootstrap/SettingsMenuContext.php:
```diff
@@ -67,12 +67,22 @@ private static function menuItemFor($itemText) {
 		$this->actor->find(self::settingsMenuButton(), 10)->click();
 	}
 
+
```
✔️

blizzz commented on tests/acceptance/features/bootstrap/SettingsMenuContext.php:
```diff
@@ -83,6 +93,14 @@ public function iOpenTheUserSettings() {
 	/**
+	 * @When I visit the settings page
+	 */
+	public function iVisitTheSettingsPage() {
```
> Besides the use of the term "page" explained above, the term "visit" is used when visiting a specific URL (although that URL may be abstracted in the step name, like in "I visit the Home page")

This is what happens

danxuliu replied:
>  This is what happens

What? :) Maybe did you forget to attach a log of the acceptance tests output or something like that? In any case, please be aware that if you change the function and annotation there you must also change it in the _tests/acceptance/features/access-levels.feature_ file too.

blizzz replied:
Well, what I quoted in my reply :) An URL is requested by clicking on the according link.

The files are fine, and all tests pass.

danxuliu replied:
>  Well, what I quoted in my reply :)

Ah, ok, sorry, I did not understand you :)

> An URL is requested by clicking on the according link.

With _the term "visit" is used when visiting a specific URL_ I meant that the URL was directly typed in the address bar and accessed. For example, `iVisitTheHomePage()` is implemented as `$this->actor->getSession()->visit($this->actor->locatePath("/"));` (which basically does what I have described). Of course you _visit_ a specific URL by clicking on a link, but I was using the term _visit_ to convey typing the URL in the address bar (ok, then maybe I should have used _iTypeTheHomePageUrlInTheAddressBar_, but I just stuck to the Mink terminology :P ). I hope that at least this is more clear than the _page_ term ;)

blizzz committed:
fix typos and unnecessary white spaces

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
> Just minor style issues from your friendly nitpicker danxuliu :P
>
> I do not want to make you work just due to my own style preferences, so I can amend the related commit myself if you agree with that ;)

Appreciated, thank you :)

blizzz committed:
We're on 13 now…

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
try to lower the timeout in an acceptance test

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

jancborchardt commented:
Just to confirm for the record: as @nickvergessen and @MorrisJobke already said, two factor auth and encryption should of course be in the security category too. :)

blizzz commented:
Two factor auth is, encryption isn't yet. I move it there tomorrow.

blizzz committed:
move encryption to security (and behind two factor things)

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
move admin templates to the right place

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
done

blizzz committed:
… and their tests needed adjustments, too

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
So, reviews? 👀

nickvergessen reviewed (approved):
Works and looks good.

I would move some sections around, but lets do that afterwards.

blizzz commented:
thx @nickvergessen and @icewind1991 🍰 

> I would move some sections around, but lets do that afterwards.

Aye. Can be an issue with a discussion and involvement with the design team.
